clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -2473,11 +2473,9 @@ bool SemaHLSL::CheckBuiltinFunctionCall(unsigned BuiltinID, CallExpr *TheCall) {
     break;
   }
   case Builtin::BI__builtin_hlsl_dot2add: {
-    // Check number of arguments should be 3
     if (SemaRef.checkArgCount(TheCall, 3))
       return true;
 
-    // Check first two arguments are vector of length 2 with half data type
     auto checkHalfVectorOfSize2 = [](clang::QualType PassedType) -> bool {
```

Comment:
So now that this isnt void(...) and you defined the `Prototype` as `float(_ExtVector<2, _Float16>,_ExtVector<2, _Float16>, float)` I don't think we need these sema checks anymore. If you delete\comment this out I would expect your sema checks to still pass.

Reply:
Makes sense. I am thinking on it a bit stretched: Now we don't need to do any semantics check (just need to set the type) because we are explicitly specifying the exact overload.
Isn't it?

Reply:
I find it very likely you won't need to do any semantic checks.  Despite what you are seeing

Look at the codegen if you are seeing something like `trunc` that is converting to half2  then this is working correctly and thats why  you aren't seeing the error you are expecting.